### PR TITLE
Check for properties that can't be found

### PR DESCRIPTION
**Pull request by rynowak**
If you give ModelExpressionProvider a lambda with a private property
you'll end up here. This wasn't common before, but it seems like users
are more likely to try it with pages.

Model Metadata and Model Binding don't handle private properties, so
supporting it in Model Expressions seems less than useful.

This isn't a breaking change because this case would have resulted in a
null-ref. Addresses #6400

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs [hunk 1]
```diff
@@ -102,6 +104,12 @@ public static class ExpressionMetadataProvider
                 //    m => m.Color.Red (nested property access)
                 //    m => m.Widgets[0].Size (expression ending with property-access)
                 metadata = metadataProvider.GetMetadataForType(containerType).Properties[propertyName];
+                if (metadata == null)
+                {
+                    throw new InvalidOperationException(Resources.FormatModelExpression_InvalidProperty(
+                        nameof(ModelMetadata),
+                        containerType.Name + "." + propertyName));
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs]: FYI there were no tests for what would happen here.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs [hunk 2]
```diff
@@ -94,6 +95,7 @@ public static class ExpressionMetadataProvider
                 //    m => foo (arbitrary expression)
                 //    m => m.Widgets[0] (expression ending with non-property-access)
                 metadata = metadataProvider.GetMetadataForType(typeof(TResult));
+                Debug.Assert(metadata != null);
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs]: We expect that this will never fail, `GetMetadataForType` always succeeds

Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs]: Then why add the assertion?

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs]: Need something better than `$"{type}.{property}"`. That isn't valid C# unless the property is `static`. It also won't mean much when the property doesn't exist at all or is a field. Suggest passing the property and type names to the resource separately.

(Personally, I'm glad we didn't have tests demonstrating an NRE! Would have been better if we'd noticed the bug of course.)

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs [hunk 3]
```diff
@@ -102,6 +104,12 @@ public static class ExpressionMetadataProvider
                 //    m => m.Color.Red (nested property access)
                 //    m => m.Widgets[0].Size (expression ending with property-access)
                 metadata = metadataProvider.GetMetadataForType(containerType).Properties[propertyName];
+                if (metadata == null)
+                {
+                    throw new InvalidOperationException(Resources.FormatModelExpression_InvalidProperty(
+                        nameof(ModelMetadata),
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs]: Don't we normally use the type's full name in exception messages like this one?

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Resources.resx [hunk 4]
```diff
@@ -292,4 +292,7 @@
   <data name="TempDataProperties_PublicGetterSetter" xml:space="preserve">
     <value>The '{0}.{1}' property with {2} is invalid. A property using {2} must have a public getter and setter.</value>
   </data>
+  <data name="ModelExpression_InvalidProperty" xml:space="preserve">
+    <value>Could not create a {0} for property '{1}'. Make sure the property is public.</value>
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Resources.resx]: `Could not create {0}. {1} must be a public instance property in '{2}'.` e.g. `Could not create Microsoft.AspNetCore.Mvc.ModelBinding.ModelMetadata. Nonexistent must be a public instance property in 'Does.Not.Contain.Nonexistent'.`

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/ExpressionMetadataProviderTest.cs [hunk 5]
```diff
@@ -148,6 +148,29 @@ public void FromStringExpression_SetsContainerAsExpected()
             Assert.Same(myModel, metadata.Container.Model);
         }
 
+        // This is a special case that's basically only possible with Razor Pages.
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/ExpressionMetadataProviderTest.cs]: Why is the `private` property case only possible with Razor Pages?

More generally, tag and HTML helpers would see similar NREs with `internal` properties that are visible in the application assembly but not to MVC, `static` properties, and indexers. &hellip; Speaking of which, please add tests abusing `static` properties and indexers.

Might also hit issues with `public` fields and perhaps method groups.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs]: It makes it clear what the expectations are on both branches wrt the null-ness of `metadata`. Zoom out and look at the whole block.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs]: We've done both, I don't have a strong preference if you think one is more correct than the other.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs]: > Need something better than $"{type}.{property}". That isn't valid C# unless the property is static. 

What's your suggestion? This doesn't need to be compilable code, it needs to tell the user what to fix.

> It also won't mean much when the property doesn't exist at all or is a field. 

You won't get here in that case. Non-existent property would be a compilation error. A field would go into the `if` and get the metadata for the field's type.

It's possible that this behavior isn't exactly right today for fields. Let's discuss.

>Suggest passing the property and type names to the resource separately.

We generally don't put things that look like code into a resource because it's easy for it be localized wrongly.  Let's discuss in person what your ideas are for improving this error message.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Resources.resx (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Resources.resx]: Does this suggestion resolve all of your other feedback around the error message?

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/ExpressionMetadataProviderTest.cs (hunk 5 above)
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/ExpressionMetadataProviderTest.cs]: > Why is the private property case only possible with Razor Pages?

You're more likely to think it would work.

I'm happy to try a few other things

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Resources.resx (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Resources.resx]: BTW that's why I don't like using full names, I think we generally have to for user types. But I would rather see `'ModelMetadata'` than `Microsoft.AspNetCore.Mvc.ModelBinding.ModelMetadata` because it's not super relevant.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs]: If this makes you sad I'm happy to take it out, I thought it was an improvement.

Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs]: After hearing your heartfelt outpouring of support for this assertion, I'm :ok: leaving this as you've got it. Just doesn't matter.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/ExpressionMetadataProvider.cs]: Talked offline: @rynowak hadn't seen the resource I suggested passing the separate property and type names into when he commented on this.

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/Resources.resx (hunk 4 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/Resources.resx]: Yes, this suggestion resolves my feedback on the error message. And I'm fine w/ `nameof(ModelMetadata)` since (as you said) it's not a user type.

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/ExpressionMetadataProviderTest.cs (hunk 5 above)
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/ExpressionMetadataProviderTest.cs]: @rynowak discussed offline and he's going to try a few more cases e.g. fields, `static` properties, and visibility-challenged properties. Indexers are handled elsewhere. Method groups probably can't be referenced in a Linq expression but @rynowak will try.

Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/ExpressionMetadataProviderTest.cs]: Aim here is to make at least the cases where `metadataProvider.GetMetadataForType(containerType).Properties[propertyName]` is called but can fail behave consistently. Secondary concern is providing reasonable error messages for the cases that will fail just after this method returns.

**Comment by rynowak** [comment]
**TLDR summary** 
So, fortunately (or unfortunately) this does not throw today for fields. It makes a best-effort and will result in rendering an `input` and `label` with the type's metadata.

If I harden this code to make a best-effort in all cases, 

```
    <form method="post">
        <div>
            // This is normal 
            <label asp-for="Message" />
            <input asp-for="Message" />
        </div>
        <div>
            // This will throw an ArgumentNullException without my changes
            <label asp-for="PrivateProperty" />
            <input asp-for="PrivateProperty" />
        </div>
        <div>
            Private
            // This will throw an NullReferenceException without my changes 
            <label asp-for="@TestModel.StaticProperty" />
            <input asp-for="@TestModel.StaticProperty" />
        </div>
        <div>
            // This will render just fine, but will use the metadata of the field's type
            <label asp-for="Field" />
            <input asp-for="Field" />
        </div>
    </form>
```

I also tried using a method group, but it fails to compile 

I suspect a static field will also throw. There's a null-ref with statics here https://github.com/aspnet/Mvc/pull/6445/files#diff-6b9d2dad7de15ba55476f18df23cb98aL63 where `memberExpression.Expression` is null. Meaning that this is a member expression with a null LHS.


------ 

@dougbu - I'm pretty strongly leaning towards making all of these things 'best effort'. If if change the cases that throw in this code today to use the metadata of the expression type, everything else works fine.

**Comment by rynowak** [comment]
@dougbu - updated this to be best effort. If you don't like this plan let's discuss tomorrow and come to a decision.

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/ExpressionMetadataProviderTest.cs [hunk 6]
```diff
@@ -159,5 +225,11 @@ private class Category
 
             public string CategoryName { get; set; }
         }
+
+        private string PrivateProperty { get; set; }
+
+        public static string StaticProperty { get; set; }
+
+        public string Field = "Hello";
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/ExpressionMetadataProviderTest.cs]: Probably better to place these at the top of the class instead of below the nested classes.

**Comment by rynowak** [comment]
As discussed in person, adding a functional test that shows that this doesn't crash when used in an end to end, and that htmlhelpers/taghelpers can retrieve the values

**Comment by rynowak** [comment]
Updated this with a functional test.

File: test/WebSites/HtmlGenerationWebSite/Views/HtmlGeneration_WeirdExpressions/GetWeirdWithHtmlHelpers.cshtml [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+        @Html.LabelFor(m => m.Field)
+        @Html.TextBoxFor(m => m.Field)
+    </div>
+    <div>
+        <p>Static Property</p>
+        @Html.LabelFor(m => WeirdModel.StaticProperty)
+        @Html.TextBoxFor(m => WeirdModel.StaticProperty)
+    </div>
+    <div>
+        <p>Field</p>
```
Comment (dougbu) [inline test/WebSites/HtmlGenerationWebSite/Views/HtmlGeneration_WeirdExpressions/GetWeirdWithHtmlHelpers.cshtml]: This isn't a field; it's **private**

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/ExpressionMetadataProviderTest.cs (hunk 6 above)
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/Internal/ExpressionMetadataProviderTest.cs]: Thx

File: test/Microsoft.AspNetCore.Mvc.FunctionalTests/HtmlGenerationTest.cs [hunk 8]
```diff
@@ -523,6 +523,30 @@ public class HtmlGenerationTest :
             Assert.Equal(expected, response, ignoreLineEndingDifferences: true);
         }
 
+        // We want to make sure that for 'wierd' model expressions involving:
+        // - fields
+        // - statics
+        // - private
+        //
+        // that we don't throw, and can evaluate the expression to get the model value. We don't
+        // expect that the model metadata returned as accurate, so we're not checking that.
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/HtmlGenerationTest.cs]: Not quite sure what this last sentence means. The fields, statics, and private property you're using have no additional metadata. So, nothing is being lost or not checked.

File: test/WebSites/HtmlGenerationWebSite/Views/HtmlGeneration_WeirdExpressions/GetWeirdWithHtmlHelpers.cshtml [hunk 9]
```diff
@@ -0,0 +1,27 @@
+@addTagHelper *, Microsoft.AspNetCore.Mvc.TagHelpers
```
Comment (dougbu) [inline test/WebSites/HtmlGenerationWebSite/Views/HtmlGeneration_WeirdExpressions/GetWeirdWithHtmlHelpers.cshtml]: Shouldn't need this here

File: test/Microsoft.AspNetCore.Mvc.FunctionalTests/HtmlGenerationTest.cs (hunk 8 above)
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/HtmlGenerationTest.cs]: >So, nothing is being lost or not checked.

Exactly, the fact that if **there were metadata attributes on those members, those metadata attributes would be ignored** is not being tested. I'll improve the description so that it's clear the test intentionally omits a detail that we decided to omit.
